Get all user batches service

## Overview 
Group managers may be allowed to create a set of users in batch (for a contest, ...). 
And should be able to list the batches of users they have already created ([see the bottom on this designs](https://france-ioi.github.io/algorea-designs/01.Groups_02.Composition_d.Invitation-modes)).

On `GET /user_batches/by-group/{group_id}/`

## Input
* group_id (in url)

## Output
* A list of the batches whose the group of group_prefix is one the input group's ancestors ~~and which are "allow_new"~~:
  - group_prefix
  - custom_prefix
  - size
  - creator_id

## Validations
* The current user is manager of the ~~input.group~~ prefix group id (or its ancestors) with at least can_manage:memberships.

> Why do we need to filter by user_batch_prefixes.allow_new?

You're right, it is a mistake. I fix it...

> > > whose the group of group_prefix is one the input group's ancestors
> > 
> > 
> > or descendants?
> 
> Don't get the idea behind this.

If we have a structure like this
```
   A
  /  \
B   C 
 \  /   \
  E     F
  / \
 G H
```
If the prefix X is given to `C`, it means that batches can create users anywhere under C, so in C, E, F, G, or H. So in this service which lists batches, we list those related to prefix X in groups C, E, F, H or H as well.
So what's written looks correct. This part is not a permission restriction, just a scope restriction.

> > The authenticated user is manager of the input.group_id (or its ancestors) with at least can_manage:memberships
> 
> In other services we require the current user (or one of his ancestor groups) to be a manager of the input group strictly (we do not consider the input group's ancestor groups). Should we really do that differently here?

In which recent services don't we do that? A manager of a group is implicitly a manager of the descendant groups.

--- a/app/api/answers/get_answers.go
+++ b/app/api/answers/get_answers.go
@@ -120,7 +120,7 @@ func (srv *Service) getAnswers(rw http.ResponseWriter, httpReq *http.Request) se
 	dataQuery, apiError := service.ApplySortingAndPaging(httpReq, dataQuery, map[string]*service.FieldSortingParams{
 		"created_at": {ColumnName: "answers.created_at", FieldType: "time"},
 		"id":         {ColumnName: "answers.id", FieldType: "int64"},
-	}, "-created_at,id", "id", false)
+	}, "-created_at,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/contests/get_administered_list.go
+++ b/app/api/contests/get_administered_list.go
@@ -108,7 +108,7 @@ func (srv *Service) getAdministeredList(w http.ResponseWriter, r *http.Request) 
 			FieldType:             "string",
 		},
 		"id": {ColumnName: "items.id", FieldType: "int64"},
-	}, "title,id", "id", false)
+	}, "title,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/contests/get_members_additional_times.go
+++ b/app/api/contests/get_members_additional_times.go
@@ -145,7 +145,7 @@ func (srv *Service) getMembersAdditionalTimes(w http.ResponseWriter, r *http.Req
 		map[string]*service.FieldSortingParams{
 			"name": {ColumnName: "found_group.name", FieldType: "string"},
 			"id":   {ColumnName: "found_group.id", FieldType: "int64"}},
-		"name,id", "id", false)
+		"name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/currentuser/get_group_invitations.go
+++ b/app/api/currentuser/get_group_invitations.go
@@ -102,7 +102,7 @@ func (srv *Service) getGroupInvitations(w http.ResponseWriter, r *http.Request) 
 		map[string]*service.FieldSortingParams{
 			"at":       {ColumnName: "group_membership_changes.at", FieldType: "time"},
 			"group_id": {ColumnName: "group_membership_changes.group_id", FieldType: "int64"}},
-		"-at,group_id", "group_id", false)
+		"-at,group_id", []string{"group_id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/currentuser/get_group_memberships.go
+++ b/app/api/currentuser/get_group_memberships.go
@@ -78,7 +78,7 @@ func (srv *Service) getGroupMemberships(w http.ResponseWriter, r *http.Request) 
 		map[string]*service.FieldSortingParams{
 			"member_since": {ColumnName: "member_since", FieldType: "time"},
 			"id":           {ColumnName: "groups.id", FieldType: "int64"}},
-		"-member_since,id", "id", false)
+		"-member_since,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/currentuser/get_group_memberships_history.go
+++ b/app/api/currentuser/get_group_memberships_history.go
@@ -76,7 +76,7 @@ func (srv *Service) getGroupMembershipsHistory(w http.ResponseWriter, r *http.Re
 		map[string]*service.FieldSortingParams{
 			"at":       {ColumnName: "group_membership_changes.at", FieldType: "time"},
 			"group_id": {ColumnName: "group_membership_changes.group_id", FieldType: "int64"}},
-		"-at,group_id", "group_id", false)
+		"-at,group_id", []string{"group_id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/currentuser/search_for_available_groups.go
+++ b/app/api/currentuser/search_for_available_groups.go
@@ -122,7 +122,7 @@ func (srv *Service) searchForAvailableGroups(w http.ResponseWriter, r *http.Requ
 	query, apiError := service.ApplySortingAndPaging(r, query,
 		map[string]*service.FieldSortingParams{
 			"id": {ColumnName: "groups.id", FieldType: "int64"}},
-		"id", "id", false)
+		"id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_children.go
+++ b/app/api/groups/get_children.go
@@ -152,7 +152,7 @@ func (srv *Service) getChildren(w http.ResponseWriter, r *http.Request) service.
 			"type":  {ColumnName: "groups.type", FieldType: "string"},
 			"grade": {ColumnName: "groups.grade", FieldType: "int64"},
 			"id":    {ColumnName: "groups.id", FieldType: "int64"}},
-		"name,id", "id", false)
+		"name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_group_progress.go
+++ b/app/api/groups/get_group_progress.go
@@ -155,7 +155,7 @@ func (srv *Service) getGroupProgress(w http.ResponseWriter, r *http.Request) ser
 		// Note that we require the 'from.name' request parameter although the service does not return group names
 		"name": {ColumnName: "group_child.name", FieldType: "string"},
 		"id":   {ColumnName: "group_child.id", FieldType: "int64"},
-	}, "name,id", "id", false)
+	}, "name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_managers.go
+++ b/app/api/groups/get_managers.go
@@ -99,7 +99,7 @@ func (srv *Service) getManagers(w http.ResponseWriter, r *http.Request) service.
 		map[string]*service.FieldSortingParams{
 			"name": {ColumnName: "groups.name"},
 			"id":   {ColumnName: "groups.id", FieldType: "int64"}},
-		"name,id", "id", false)
+		"name,id", []string{"id"}, false)
 
 	if apiError != service.NoError {
 		return apiError

--- a/app/api/groups/get_members.go
+++ b/app/api/groups/get_members.go
@@ -149,7 +149,7 @@ func (srv *Service) getMembers(w http.ResponseWriter, r *http.Request) service.A
 			"user.grade":   {ColumnName: "users.grade"},
 			"member_since": {ColumnName: "member_since", FieldType: "time"},
 			"id":           {ColumnName: "groups_groups.child_group_id", FieldType: "int64"}},
-		"-member_since,id", "id", false)
+		"-member_since,id", []string{"id"}, false)
 
 	if apiError != service.NoError {
 		return apiError

--- a/app/api/groups/get_requests.go
+++ b/app/api/groups/get_requests.go
@@ -202,7 +202,7 @@ func (srv *Service) getRequests(w http.ResponseWriter, r *http.Request) service.
 			"joining_user.login": {ColumnName: "joining_user.login"},
 			"at":                 {ColumnName: "group_membership_changes.at", FieldType: "time"},
 			"member_id":          {ColumnName: "group_membership_changes.member_id", FieldType: "int64"}},
-		"-at,member_id", "member_id", false)
+		"-at,member_id", []string{"member_id"}, false)
 
 	if apiError != service.NoError {
 		return apiError

--- a/app/api/groups/get_team_descendants.go
+++ b/app/api/groups/get_team_descendants.go
@@ -81,7 +81,7 @@ func (srv *Service) getTeamDescendants(w http.ResponseWriter, r *http.Request) s
 		map[string]*service.FieldSortingParams{
 			"name": {ColumnName: "groups.name", FieldType: "string"},
 			"id":   {ColumnName: "groups.id", FieldType: "int64"}},
-		"name,id", "id", false)
+		"name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_team_progress.go
+++ b/app/api/groups/get_team_progress.go
@@ -128,7 +128,7 @@ func (srv *Service) getTeamProgress(w http.ResponseWriter, r *http.Request) serv
 		// Note that we require the 'from.name' request parameter although the service does not return group names
 		"name": {ColumnName: "groups.name", FieldType: "string"},
 		"id":   {ColumnName: "groups.id", FieldType: "int64"},
-	}, "name,id", "id", false)
+	}, "name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_user_batches.feature
+++ b/app/api/groups/get_user_batches.feature
@@ -40,7 +40,7 @@ Feature: List user batches (userBatchesView)
 
   Scenario: List user batches (default sort)
     Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/13"
+    When I send a GET request to "/user-batches/by-group/21"
     Then the response code should be 200
     And the response body should be, in JSON:
     """
@@ -57,7 +57,7 @@ Feature: List user batches (userBatchesView)
 
   Scenario: List user batches (sorted by group_prefix desc, size desc)
     Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/13?sort=-group_prefix,-size"
+    When I send a GET request to "/user-batches/by-group/21?sort=-group_prefix,-size"
     Then the response code should be 200
     And the response body should be, in JSON:
     """
@@ -74,7 +74,7 @@ Feature: List user batches (userBatchesView)
 
   Scenario: List user batches (sorted by custom_prefix desc)
     Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/13?sort=-custom_prefix"
+    When I send a GET request to "/user-batches/by-group/21?sort=-custom_prefix"
     Then the response code should be 200
     And the response body should be, in JSON:
     """
@@ -91,7 +91,7 @@ Feature: List user batches (userBatchesView)
 
   Scenario: List user batches (sorted by size)
     Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/13?sort=size"
+    When I send a GET request to "/user-batches/by-group/21?sort=size"
     Then the response code should be 200
     And the response body should be, in JSON:
     """
@@ -108,7 +108,7 @@ Feature: List user batches (userBatchesView)
 
   Scenario: List user batches (sorted by size, start from the second row)
     Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/13?sort=size&from.size=200&from.group_prefix=test&from.custom_prefix=custom"
+    When I send a GET request to "/user-batches/by-group/21?sort=size&from.size=200&from.group_prefix=test&from.custom_prefix=custom"
     Then the response code should be 200
     And the response body should be, in JSON:
     """

--- a/app/api/groups/get_user_batches.feature
+++ b/app/api/groups/get_user_batches.feature
@@ -1,0 +1,123 @@
+Feature: List user batches (userBatchesView)
+  Background:
+    Given the database has the following table 'groups':
+      | id | name   | grade | type  |
+      | 13 | class  | -2    | Class |
+      | 14 | class2 | -2    | Class |
+      | 21 | user   | -2    | User  |
+    And the database has the following table 'users':
+      | login | group_id |
+      | owner | 21       |
+    And the database has the following table 'group_managers':
+      | group_id | manager_id | can_manage  |
+      | 13       | 21         | memberships |
+    And the database has the following table 'groups_ancestors':
+      | ancestor_group_id | child_group_id |
+      | 13                | 13             |
+      | 13                | 21             |
+      | 14                | 14             |
+      | 21                | 21             |
+    And the database has the following table 'groups_groups':
+      | parent_group_id | child_group_id |
+      | 13              | 21             |
+    And the database has the following table 'user_batch_prefixes':
+      | group_prefix | group_id | allow_new |
+      | test         | 13       | 1         |
+      | test1        | 13       | 1         |
+      | test2        | 13       | 0         |
+      | test3        | 21       | 1         |
+      | test4        | 14       | 1         |
+    And the database has the following table 'user_batches':
+      | group_prefix | custom_prefix | size | creator_id |
+      | test         | custom        | 100  | null       |
+      | test         | custom1       | 200  | 13         |
+      | test1        | cust          | 300  | 21         |
+      | test1        | cust1         | 400  | null       |
+      | test2        | cus           | 300  | 21         |
+      | test2        | cus1          | 400  | null       |
+      | test3        | pref          | 500  | null       |
+      | test4        | prf           | 600  | null       |
+
+  Scenario: List user batches (default sort)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": null, "custom_prefix": "custom", "group_prefix": "test", "size": 100},
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400},
+      {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500}
+    ]
+    """
+
+  Scenario: List user batches (sorted by group_prefix desc, size desc)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=-group_prefix,-size"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": null, "custom_prefix": "custom", "group_prefix": "test", "size": 100}
+    ]
+    """
+
+  Scenario: List user batches (sorted by custom_prefix desc)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=-custom_prefix"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500},
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": null, "custom_prefix": "custom", "group_prefix": "test", "size": 100},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300}
+    ]
+    """
+
+  Scenario: List user batches (sorted by size)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=size"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": null, "custom_prefix": "custom", "group_prefix": "test", "size": 100},
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400},
+      {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500}
+    ]
+    """
+
+  Scenario: List user batches (sorted by size, start from the second row)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=size&from.size=200&from.group_prefix=test&from.custom_prefix=custom"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400},
+      {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500}
+    ]
+    """

--- a/app/api/groups/get_user_batches.feature
+++ b/app/api/groups/get_user_batches.feature
@@ -121,3 +121,28 @@ Feature: List user batches (userBatchesView)
       {"creator_id": null, "custom_prefix": "pref", "group_prefix": "test3", "size": 500}
     ]
     """
+
+  Scenario: List user batches (default sort, another group_id)
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+      {"creator_id": null, "custom_prefix": "custom", "group_prefix": "test", "size": 100},
+      {"creator_id": 13, "custom_prefix": "custom1", "group_prefix": "test", "size": 200},
+      {"creator_id": 21, "custom_prefix": "cust", "group_prefix": "test1", "size": 300},
+      {"creator_id": null, "custom_prefix": "cust1", "group_prefix": "test1", "size": 400},
+      {"creator_id": 21, "custom_prefix": "cus", "group_prefix": "test2", "size": 300},
+      {"creator_id": null, "custom_prefix": "cus1", "group_prefix": "test2", "size": 400}
+    ]
+    """
+
+  Scenario: User is not a manager of the group_id
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/14"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    []
+    """

--- a/app/api/groups/get_user_batches.go
+++ b/app/api/groups/get_user_batches.go
@@ -26,14 +26,10 @@ type userBatch struct {
 // summary: List user batches
 // description: >
 //
-//   Lists the batches of users whose prefix can be used in the given group (i.e., the `group_id` is a descendant of the prefix group).
-//
-//
-//   Restrictions:
-//
-//   * The authenticated user (or one of his group ancestors) should be a manager of the `group_id`
-//     (directly, or of one of its ancestors)
-//     with at least 'can_manage:memberships', otherwise the 'forbidden' response is returned.
+//   Lists the batches of users whose prefix can be used in the given group
+//   (i.e., the `group_id` is a descendant of the prefix group).
+//   Only those user batches are shown for which the authenticated user (or one of his group ancestors) is a manager of
+//   the prefix group (or its ancestor) with at least 'can_manage:memberships'.
 // parameters:
 // - name: group_id
 //   in: path
@@ -90,20 +86,17 @@ func (srv *Service) getUserBatches(w http.ResponseWriter, r *http.Request) servi
 		return service.ErrInvalidRequest(err)
 	}
 
-	found, err := srv.Store.ActiveGroupAncestors().ManagedByUser(user).
-		Where("groups_ancestors_active.child_group_id = ?", groupID).
-		Where("can_manage != 'none'").HasRows()
-	service.MustNotBeError(err)
-	if !found {
-		return service.InsufficientAccessRightsError
-	}
+	managedByUser := srv.Store.ActiveGroupAncestors().ManagedByUser(user).
+		Where("can_manage != 'none'").
+		Select("groups_ancestors_active.child_group_id AS id")
+
+	prefixAncestors := srv.Store.ActiveGroupAncestors().Where("child_group_id = ?", groupID).
+		Select("ancestor_group_id AS id")
 
 	query := srv.Store.UserBatches().
 		Joins("JOIN user_batch_prefixes USING(group_prefix)").
-		Joins(`
-			JOIN groups_ancestors_active
-				ON groups_ancestors_active.ancestor_group_id = user_batch_prefixes.group_id AND
-				   groups_ancestors_active.child_group_id = ?`, groupID).
+		Where(`user_batch_prefixes.group_id IN(?)`, managedByUser.QueryExpr()).
+		Where(`user_batch_prefixes.group_id IN(?)`, prefixAncestors.QueryExpr()).
 		Select("group_prefix, custom_prefix, size, creator_id")
 
 	query, apiErr := service.ApplySortingAndPaging(r, query, map[string]*service.FieldSortingParams{

--- a/app/api/groups/get_user_batches.go
+++ b/app/api/groups/get_user_batches.go
@@ -26,7 +26,7 @@ type userBatch struct {
 // summary: List user batches
 // description: >
 //
-//   Lists existing batches of users.
+//   Lists existing batches of users whose group_prefix's `group_id` is one the input `group_id`'s descendants.
 //
 //
 //   Restrictions:

--- a/app/api/groups/get_user_batches.go
+++ b/app/api/groups/get_user_batches.go
@@ -26,7 +26,7 @@ type userBatch struct {
 // summary: List user batches
 // description: >
 //
-//   Lists existing batches of users whose group_prefix's `group_id` is one the input `group_id`'s descendants.
+//   Lists the batches of users whose prefix can be used in the given group (i.e., the `group_id` is a descendant of the prefix group).
 //
 //
 //   Restrictions:
@@ -102,8 +102,8 @@ func (srv *Service) getUserBatches(w http.ResponseWriter, r *http.Request) servi
 		Joins("JOIN user_batch_prefixes USING(group_prefix)").
 		Joins(`
 			JOIN groups_ancestors_active
-				ON groups_ancestors_active.ancestor_group_id = ? AND
-				   groups_ancestors_active.child_group_id = user_batch_prefixes.group_id`, groupID).
+				ON groups_ancestors_active.ancestor_group_id = user_batch_prefixes.group_id AND
+				   groups_ancestors_active.child_group_id = ?`, groupID).
 		Select("group_prefix, custom_prefix, size, creator_id")
 
 	query, apiErr := service.ApplySortingAndPaging(r, query, map[string]*service.FieldSortingParams{

--- a/app/api/groups/get_user_batches.go
+++ b/app/api/groups/get_user_batches.go
@@ -58,7 +58,7 @@ type userBatch struct {
 //   in: query
 //   type: string
 // - name: limit
-//   description: Display the first N members
+//   description: Display the first N user batches
 //   in: query
 //   type: integer
 //   maximum: 1000

--- a/app/api/groups/get_user_batches.go
+++ b/app/api/groups/get_user_batches.go
@@ -1,0 +1,123 @@
+package groups
+
+import (
+	"net/http"
+
+	"github.com/go-chi/render"
+
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+)
+
+// swagger:model userBatch
+type userBatch struct {
+	// required: true
+	GroupPrefix string `json:"group_prefix"`
+	// required: true
+	CustomPrefix string `json:"custom_prefix"`
+	// required: true
+	Size int `json:"size"`
+	// Nullable
+	// required: true
+	CreatorID *int64 `json:"creator_id"`
+}
+
+// swagger:operation GET /user-batches/by-group/{group_id} groups userBatchesView
+// ---
+// summary: List user batches
+// description: >
+//
+//   Lists existing batches of users.
+//
+//
+//   Restrictions:
+//
+//   * The authenticated user (or one of his group ancestors) should be a manager of the `group_id`
+//     (directly, or of one of its ancestors)
+//     with at least 'can_manage:memberships', otherwise the 'forbidden' response is returned.
+// parameters:
+// - name: group_id
+//   in: path
+//   required: true
+//   type: integer
+// - name: sort
+//   in: query
+//   default: [group_prefix,custom_prefix]
+//   type: array
+//   items:
+//     type: string
+//     enum: [group_prefix,-group_prefix,custom_prefix,-custom_prefix,size,-size]
+// - name: from.group_prefix
+//   description: Start the page from the batch next to the batch with `user_batches.group_prefix` = `from.group_prefix`
+//                (`from.custom_prefix` is required when `from.group_prefix` is given)
+//   in: query
+//   type: string
+// - name: from.custom_prefix
+//   description: Start the page from the batch next to the batch with `user_batches.custom_prefix` = `from.custom_prefix`
+//                (`from.group_prefix` is required when `from.custom_prefix` is given)
+//   in: query
+//   type: string
+// - name: from.size
+//   description: Start the page from the batch next to the batch with `user_batches.size` = `from.size`
+//                (`from.group_prefix` & `from.custom_prefix` are required when `from.size` is given)
+//   in: query
+//   type: string
+// - name: limit
+//   description: Display the first N members
+//   in: query
+//   type: integer
+//   maximum: 1000
+//   default: 500
+// responses:
+//   "200":
+//     description: OK. The array of user batches
+//     schema:
+//       type: array
+//       items:
+//         "$ref": "#/definitions/userBatch"
+//   "400":
+//     "$ref": "#/responses/badRequestResponse"
+//   "401":
+//     "$ref": "#/responses/unauthorizedResponse"
+//   "403":
+//     "$ref": "#/responses/forbiddenResponse"
+//   "500":
+//     "$ref": "#/responses/internalErrorResponse"
+func (srv *Service) getUserBatches(w http.ResponseWriter, r *http.Request) service.APIError {
+	user := srv.GetUser(r)
+
+	groupID, err := service.ResolveURLQueryPathInt64Field(r, "group_id")
+	if err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	found, err := srv.Store.ActiveGroupAncestors().ManagedByUser(user).
+		Where("groups_ancestors_active.child_group_id = ?", groupID).
+		Where("can_manage != 'none'").HasRows()
+	service.MustNotBeError(err)
+	if !found {
+		return service.InsufficientAccessRightsError
+	}
+
+	query := srv.Store.UserBatches().
+		Joins("JOIN user_batch_prefixes USING(group_prefix)").
+		Joins(`
+			JOIN groups_ancestors_active
+				ON groups_ancestors_active.ancestor_group_id = ? AND
+				   groups_ancestors_active.child_group_id = user_batch_prefixes.group_id`, groupID).
+		Select("group_prefix, custom_prefix, size, creator_id")
+
+	query, apiErr := service.ApplySortingAndPaging(r, query, map[string]*service.FieldSortingParams{
+		"group_prefix":  {ColumnName: "group_prefix", FieldType: "string"},
+		"custom_prefix": {ColumnName: "custom_prefix", FieldType: "string"},
+		"size":          {ColumnName: "size", FieldType: "int64"},
+	}, "group_prefix,custom_prefix", []string{"group_prefix", "custom_prefix"}, false)
+	if apiErr != service.NoError {
+		return apiErr
+	}
+
+	var result []userBatch
+	service.MustNotBeError(query.Scan(&result).Error())
+
+	render.Respond(w, r, result)
+	return service.NoError
+}

--- a/app/api/groups/get_user_batches.robustness.feature
+++ b/app/api/groups/get_user_batches.robustness.feature
@@ -1,0 +1,47 @@
+Feature: List user batches (userBatchesView) - robustness
+  Background:
+    Given the database has the following table 'groups':
+      | id | name   | grade | type  |
+      | 13 | class  | -2    | Class |
+      | 14 | class2 | -2    | Class |
+      | 21 | user   | -2    | User  |
+    And the database has the following table 'users':
+      | login | group_id |
+      | owner | 21       |
+    And the database has the following table 'group_managers':
+      | group_id | manager_id | can_manage  |
+      | 13       | 21         | memberships |
+      | 14       | 21         | none        |
+    And the database has the following table 'groups_ancestors':
+      | ancestor_group_id | child_group_id |
+      | 13                | 13             |
+      | 13                | 21             |
+      | 14                | 14             |
+      | 21                | 21             |
+    And the database has the following table 'groups_groups':
+      | parent_group_id | child_group_id |
+      | 13              | 21             |
+
+  Scenario: Invalid group_id given
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/1_1"
+    Then the response code should be 400
+    And the response error message should contain "Wrong value for group_id (should be int64)"
+
+  Scenario: User is not a manager of the group_id
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/14"
+    Then the response code should be 403
+    And the response error message should contain "Insufficient access rights"
+
+  Scenario: Invalid sorting rules given
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=code"
+    Then the response code should be 400
+    And the response error message should contain "Unallowed field in sorting parameters: "code""
+
+  Scenario: A tie-breaker field is missing
+    Given I am the user with id "21"
+    When I send a GET request to "/user-batches/by-group/13?sort=size&from.size=200"
+    Then the response code should be 400
+    And the response error message should contain "All 'from' parameters (from.size, from.group_prefix, from.custom_prefix) or none of them must be present"

--- a/app/api/groups/get_user_batches.robustness.feature
+++ b/app/api/groups/get_user_batches.robustness.feature
@@ -28,12 +28,6 @@ Feature: List user batches (userBatchesView) - robustness
     Then the response code should be 400
     And the response error message should contain "Wrong value for group_id (should be int64)"
 
-  Scenario: User is not a manager of the group_id
-    Given I am the user with id "21"
-    When I send a GET request to "/user-batches/by-group/14"
-    Then the response code should be 403
-    And the response error message should contain "Insufficient access rights"
-
   Scenario: Invalid sorting rules given
     Given I am the user with id "21"
     When I send a GET request to "/user-batches/by-group/13?sort=code"

--- a/app/api/groups/get_user_descendants.go
+++ b/app/api/groups/get_user_descendants.go
@@ -84,7 +84,7 @@ func (srv *Service) getUserDescendants(w http.ResponseWriter, r *http.Request) s
 		map[string]*service.FieldSortingParams{
 			"name": {ColumnName: "groups.name", FieldType: "string"},
 			"id":   {ColumnName: "groups.id", FieldType: "int64"}},
-		"name,id", "id", false)
+		"name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/get_user_progress.go
+++ b/app/api/groups/get_user_progress.go
@@ -133,7 +133,7 @@ func (srv *Service) getUserProgress(w http.ResponseWriter, r *http.Request) serv
 		// Note that we require the 'from.name' request parameter although the service does not return group names
 		"name": {ColumnName: "groups.name", FieldType: "string"},
 		"id":   {ColumnName: "groups.id", FieldType: "int64"},
-	}, "name,id", "id", false)
+	}, "name,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/groups/groups.go
+++ b/app/api/groups/groups.go
@@ -58,6 +58,7 @@ func (srv *Service) SetRoutes(router chi.Router) {
 
 	router.Get("/current-user/teams/by-item/{item_id}", service.AppHandler(srv.getCurrentUserTeamByItem).ServeHTTP)
 	router.Post("/user-batches", service.AppHandler(srv.createUserBatch).ServeHTTP)
+	router.Get("/user-batches/by-group/{group_id}", service.AppHandler(srv.getUserBatches).ServeHTTP)
 }
 
 func checkThatUserCanManageTheGroup(store *database.DataStore, user *database.User, groupID int64) service.APIError {

--- a/app/api/groups/recent_activity.go
+++ b/app/api/groups/recent_activity.go
@@ -159,7 +159,7 @@ func (srv *Service) getRecentActivity(w http.ResponseWriter, r *http.Request) se
 		map[string]*service.FieldSortingParams{
 			"created_at": {ColumnName: "answers.created_at", FieldType: "time"},
 			"id":         {ColumnName: "answers.id", FieldType: "int64"}},
-		"-created_at,id", "id", false)
+		"-created_at,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/api/items/get_attempts.go
+++ b/app/api/items/get_attempts.go
@@ -141,7 +141,7 @@ func (srv *Service) getAttempts(w http.ResponseWriter, r *http.Request) service.
 	query, apiError := service.ApplySortingAndPaging(r, query, map[string]*service.FieldSortingParams{
 		"order": {ColumnName: "attempts.order", FieldType: "int64"},
 		"id":    {ColumnName: "attempts.id", FieldType: "int64"},
-	}, "order,id", "id", false)
+	}, "order,id", []string{"id"}, false)
 	if apiError != service.NoError {
 		return apiError
 	}

--- a/app/service/sorting.go
+++ b/app/service/sorting.go
@@ -48,10 +48,10 @@ func (d sortingDirection) conditionSign() string {
 // taking into the account the URL parameters 'from.*'.
 // When the `skipSortParameter` is true, the 'sort' request parameter is ignored.
 func ApplySortingAndPaging(r *http.Request, query *database.DB, acceptedFields map[string]*FieldSortingParams,
-	defaultRules, tieBreakerFieldName string, skipSortParameter bool) (*database.DB, APIError) {
+	defaultRules string, tieBreakerFieldNames []string, skipSortParameter bool) (*database.DB, APIError) {
 	sortingRules := prepareSortingRulesAndAcceptedFields(r, defaultRules, skipSortParameter)
 
-	usedFields, fieldsDirections, err := parseSortingRules(sortingRules, acceptedFields, tieBreakerFieldName)
+	usedFields, fieldsDirections, err := parseSortingRules(sortingRules, acceptedFields, tieBreakerFieldNames)
 	if err != nil {
 		return nil, ErrInvalidRequest(err)
 	}
@@ -80,7 +80,7 @@ func prepareSortingRulesAndAcceptedFields(r *http.Request, defaultRules string, 
 
 // parseSortingRules returns a slice with used fields and a map fieldName -> direction
 // It also checks that there are no unallowed fields in the rules.
-func parseSortingRules(sortingRules string, acceptedFields map[string]*FieldSortingParams, tieBreakerFieldName string) (
+func parseSortingRules(sortingRules string, acceptedFields map[string]*FieldSortingParams, tieBreakerFieldNames []string) (
 	usedFields []string, fieldsDirections map[string]sortingDirection, err error) {
 	sortStatements := strings.Split(sortingRules, ",")
 	usedFields = make([]string, 0, len(sortStatements)+1)
@@ -101,9 +101,11 @@ func parseSortingRules(sortingRules string, acceptedFields map[string]*FieldSort
 		}
 	}
 	if !includesUniqueField {
-		if fieldsDirections[tieBreakerFieldName] == 0 {
-			fieldsDirections[tieBreakerFieldName] = 1
-			usedFields = append(usedFields, tieBreakerFieldName)
+		for _, tieBreakerFieldName := range tieBreakerFieldNames {
+			if fieldsDirections[tieBreakerFieldName] == 0 {
+				fieldsDirections[tieBreakerFieldName] = 1
+				usedFields = append(usedFields, tieBreakerFieldName)
+			}
 		}
 	}
 	return

--- a/app/service/sorting_test.go
+++ b/app/service/sorting_test.go
@@ -17,11 +17,11 @@ import (
 
 func TestApplySorting(t *testing.T) {
 	type args struct {
-		urlParameters       string
-		acceptedFields      map[string]*FieldSortingParams
-		defaultRules        string
-		tieBreakerFieldName string
-		skipSortParameters  bool
+		urlParameters        string
+		acceptedFields       map[string]*FieldSortingParams
+		defaultRules         string
+		tieBreakerFieldNames []string
+		skipSortParameters   bool
 	}
 	tests := []struct {
 		name             string
@@ -38,8 +38,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL:      "SELECT id FROM `users` ORDER BY name DESC, id ASC",
 			wantAPIError: NoError},
@@ -50,7 +50,7 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				tieBreakerFieldName: "id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL:      "SELECT id FROM `users` ORDER BY name ASC, id DESC",
 			wantAPIError: NoError},
@@ -61,9 +61,9 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				tieBreakerFieldName: "id",
-				skipSortParameters:  true,
-				defaultRules:        "id",
+				tieBreakerFieldNames: []string{"id"},
+				skipSortParameters:   true,
+				defaultRules:         "id",
 			},
 			wantSQL:      "SELECT id FROM `users` ORDER BY id ASC",
 			wantAPIError: NoError},
@@ -74,8 +74,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string", ColumnNameForOrdering: "LOWER(name)"},
 					"id":   {ColumnName: "id", FieldType: "int64", ColumnNameForOrdering: "-id"},
 				},
-				tieBreakerFieldName: "id",
-				defaultRules:        "-name,id",
+				tieBreakerFieldNames: []string{"id"},
+				defaultRules:         "-name,id",
 			},
 			wantSQL:      "SELECT id FROM `users` ORDER BY LOWER(name) DESC, -id ASC",
 			wantAPIError: NoError},
@@ -86,8 +86,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantAPIError: ErrInvalidRequest(errors.New(`a field cannot be a sorting parameter more than once: "name"`))},
 		{name: "unallowed field",
@@ -97,8 +97,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantAPIError: ErrInvalidRequest(errors.New(`unallowed field in sorting parameters: "class"`))},
 		{name: "add a tie-breaker field",
@@ -108,10 +108,38 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL:          "SELECT id FROM `users` ORDER BY name DESC, id ASC",
+			wantSQLArguments: nil,
+			wantAPIError:     NoError},
+		{name: "add multiple tie-breaker fields",
+			args: args{
+				urlParameters: "",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name":     {ColumnName: "name", FieldType: "string"},
+					"group_id": {ColumnName: "group_id", FieldType: "int64"},
+					"item_id":  {ColumnName: "item_id", FieldType: "int64"},
+				},
+				defaultRules:         "-name",
+				tieBreakerFieldNames: []string{"group_id", "item_id"},
+			},
+			wantSQL:          "SELECT id FROM `users` ORDER BY name DESC, group_id ASC, item_id ASC",
+			wantSQLArguments: nil,
+			wantAPIError:     NoError},
+		{name: "add some of tie-breaker fields",
+			args: args{
+				urlParameters: "",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name":     {ColumnName: "name", FieldType: "string"},
+					"group_id": {ColumnName: "group_id", FieldType: "int64"},
+					"item_id":  {ColumnName: "item_id", FieldType: "int64"},
+				},
+				defaultRules:         "-name,item_id",
+				tieBreakerFieldNames: []string{"group_id", "item_id"},
+			},
+			wantSQL:          "SELECT id FROM `users` ORDER BY name DESC, item_id ASC, group_id ASC",
 			wantSQLArguments: nil,
 			wantAPIError:     NoError},
 		{name: "do not add a tie-breaker field if a sorting field is unique",
@@ -121,8 +149,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string", Unique: true},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL:          "SELECT id FROM `users` ORDER BY name DESC",
 			wantSQLArguments: nil,
@@ -135,8 +163,8 @@ func TestApplySorting(t *testing.T) {
 					"id":   {ColumnName: "id", FieldType: "int64"},
 					"flag": {ColumnName: "bFlag", FieldType: "bool"},
 				},
-				defaultRules:        "-name,id,flag",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id,flag",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL: "SELECT id FROM `users` " +
 				"WHERE ((name < ?) OR (name = ? AND id > ?) OR (name = ? AND id = ? AND bFlag > ?)) " +
@@ -150,8 +178,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantAPIError: ErrInvalidRequest(errors.New(`wrong value for from.id (should be int64)`))},
 		{name: "one of the from. fields is skipped",
@@ -162,8 +190,8 @@ func TestApplySorting(t *testing.T) {
 					"type": {ColumnName: "type", FieldType: "string"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantAPIError: ErrInvalidRequest(errors.New(`all 'from' parameters (from.name, from.id) or none of them must be present`))},
 		{name: "unsupported field type",
@@ -173,8 +201,8 @@ func TestApplySorting(t *testing.T) {
 					"name": {ColumnName: "name", FieldType: "interface{}"},
 					"id":   {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "-name,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "-name,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			shouldPanic: errors.New(`unsupported type "interface{}" for field "name"`)},
 		{name: "unallowed from fields",
@@ -184,8 +212,8 @@ func TestApplySorting(t *testing.T) {
 					"id":   {ColumnName: "id", FieldType: "int64"},
 					"name": {ColumnName: "name", FieldType: "string"},
 				},
-				defaultRules:        "id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantAPIError: ErrInvalidRequest(errors.New(`unallowed paging parameters (from.field, from.name, from.version)`))},
 		{name: "paging by time",
@@ -195,8 +223,8 @@ func TestApplySorting(t *testing.T) {
 					"submitted_at": {ColumnName: "submitted_at", FieldType: "time"},
 					"id":           {ColumnName: "id", FieldType: "int64"},
 				},
-				defaultRules:        "submitted_at,id",
-				tieBreakerFieldName: "id",
+				defaultRules:         "submitted_at,id",
+				tieBreakerFieldNames: []string{"id"},
 			},
 			wantSQL: "SELECT id FROM `users`  WHERE ((submitted_at > ?) OR (submitted_at = ? AND id > ?)) " +
 				"ORDER BY submitted_at ASC, id ASC",
@@ -232,7 +260,7 @@ func TestApplySorting(t *testing.T) {
 			query := db.Table("users").Select("id")
 
 			query, gotAPIError := ApplySortingAndPaging(request, query, tt.args.acceptedFields, tt.args.defaultRules,
-				tt.args.tieBreakerFieldName, tt.args.skipSortParameters)
+				tt.args.tieBreakerFieldNames, tt.args.skipSortParameters)
 			assert.Equal(t, tt.wantAPIError, gotAPIError)
 
 			if gotAPIError == NoError {

